controllers/etcdcluster_controller.go

```diff
@@ -0,0 +1,84 @@
+package controllers
+
+import (
+	"context"
+	"time"
+
+	"github.com/go-logr/logr"
+	v1 "k8s.io/api/core/v1"
+	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	ctrl "sigs.k8s.io/controller-runtime"
+	"sigs.k8s.io/controller-runtime/pkg/client"
+
+	etcdv1alpha1 "github.com/improbable-eng/etcd-cluster-operator/api/v1alpha1"
+)
+
+// EtcdClusterReconciler reconciles a EtcdCluster object
+type EtcdClusterReconciler struct {
+	client.Client
+	Log logr.Logger
+}
+
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters,verbs=get;list;watch
+// +kubebuilder:rbac:groups=etcd.improbable.io,resources=etcdclusters/status,verbs=get;update;patch
+// +kubebuilder:rbac:groups=core,resources=services,verbs=get;list;watch;create
+
+func (r *EtcdClusterReconciler) Reconcile(req ctrl.Request) (ctrl.Result, error) {
+	ctx, cancel := context.WithTimeout(context.Background(), 10*time.Second)
+	defer cancel()
+
+	log := r.Log.WithValues("etcdcluster", req.NamespacedName)
+
+	var cluster etcdv1alpha1.EtcdCluster
+	if err := r.Get(ctx, req.NamespacedName, &cluster); err != nil {
+		log.Error(err, "unable to fetch EtcdCluster")
+		return ctrl.Result{}, client.IgnoreNotFound(err)
+	}
+
+	service := &v1.Service{
+		ObjectMeta: metav1.ObjectMeta{
+			Name:      cluster.Name,
+			Namespace: cluster.Namespace,
+			OwnerReferences: []metav1.OwnerReference{
+				*metav1.NewControllerRef(&cluster, etcdv1alpha1.GroupVersion.WithKind("EtcdCluster")),
+			},
+			Labels: map[string]string{
+				appLabel:     appName,
+				clusterLabel: cluster.Name,
+			},
+		},
+		Spec: v1.ServiceSpec{
+			ClusterIP:                v1.ClusterIPNone,
+			PublishNotReadyAddresses: true,
```

Comment:
could this be a problem if we're using the service for client access? we don't want to route client traffic to pods until they're ready

Reply:
Unfortunately we *must* have this here otherwise if we consider a bootstrapping pod to be unready (when we require a working cluster in a readiness probe - which admitally we don't configure yet) then the pods will never be able to find each other.

We could advise users to create their own service for client access if their clients cannot handle unready pods?

Reply:
Yep - this is a good reason to not utilise the operator-provided service.

That said, the gRPC client also performs health checks as far as I know, so it needn't be a huge issue I think?